pkg/submitter/submitter.go

```diff
@@ -0,0 +1,504 @@
+	proxy            db.DataServerProxy
+	account          *rpc.Account
+	client           contracts.ETHClient
+	tellor           *contracts.ITellor
+	currentChallenge *mining.MiningChallenge
+	currentNonce     string
+	currentValues    [5]*big.Int
+	submitter        tellorCommon.TransactionSubmitter
+	resultCh         chan *mining.Result
+	submitCount      prometheus.Counter
+	submitFailCount  prometheus.Counter
+	submitProfit     *prometheus.GaugeVec
+	submitCost       *prometheus.GaugeVec
+	submitReward     *prometheus.GaugeVec
+	lastSubmitCncl   context.CancelFunc
+}
+
+func CreateSubmitter(ctx context.Context, cfg *config.Config, logger log.Logger, client contracts.ETHClient, tellor *contracts.ITellor, account *rpc.Account, txSubmitter tellorCommon.TransactionSubmitter, proxy db.DataServerProxy) (*Submitter, chan *mining.Result) {
+	ctx, close := context.WithCancel(ctx)
+	submitter := &Submitter{
+		ctx:       ctx,
+		close:     close,
+		proxy:     proxy,
+		submitter: txSubmitter,
+		cfg:       cfg,
+		resultCh:  make(chan *mining.Result),
+		account:   account,
+		logger:    logger,
+		tellor:    tellor,
+		submitCount: promauto.NewCounter(prometheus.CounterOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_total",
+			Help:        "The total number of submitted solutions",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		}),
+		submitFailCount: promauto.NewCounter(prometheus.CounterOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_fails_total",
+			Help:        "The total number of failed submission",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		}),
+		submitProfit: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_profit",
+			Help:        "The current submit profit in percents",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+		submitCost: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_cost",
+			Help:        "The current submit cost in 1e18 eth",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+		submitReward: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_reward",
+			Help:        "The current reward in 1e18 eth",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+	}
+
+	return submitter, submitter.resultCh
+}
+
+func (s *Submitter) Start() error {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return nil
+		case result := <-s.resultCh:
+			if s.lastSubmitCncl != nil {
+				s.lastSubmitCncl()
+			}
+			var ctx context.Context
+			ctx, s.lastSubmitCncl = context.WithCancel(s.ctx)
+			s.handleSubmit(ctx, result)
```

Comment:
I think should have it here since there is no need to run any other checks untill it is time to submit.
```suggestion
			s.blockUntilTimeToSubmit(s.ctx)
			s.handleSubmit(ctx, result)
```